apps/barista-backend/src/search/search.service.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * @license
+ * Copyright 2020 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Injectable, HttpService, Logger } from '@nestjs/common';
+import { environment } from '../environments/environment';
+import { Index } from 'lunr';
+import {
+  BaSearchResult,
+  BaPageBuilderContentResult,
+} from '@dynatrace/barista-components/barista-definitions';
+import { Observable, of, combineLatest } from 'rxjs';
+import { tap, map, take } from 'rxjs/operators';
+import { mapContentToSearchResult } from './map-content-to-search-result';
+
+// 5 Minutes of cache.
+const CACHE_EXPIRES_DURATION = 1000 * 60 * 5;
+const RESULT_THRESHOLD = 2;
+
+@Injectable()
+export class SearchService {
+  private _index: Index;
+
+  private _documentCache: Map<
+    string,
+    { timestamp: number; data: BaPageBuilderContentResult }
+  > = new Map();
+
+  constructor(private readonly _http: HttpService) {
+    this._refreshSearchIndex();
+  }
+
+  /**
+   * Finds something in the search index based on the passed
+   * query parameter and returns a list of search results.
+   * @param query Query string that should be searched
+   */
+  find(query: string): Observable<BaSearchResult[]> {
+    if (this._index) {
+      const results = this._index.search(query);
+
+      const allPages = results
+        .filter(result => result.score > RESULT_THRESHOLD)
+        .map(r => this._getContentSource(r.ref));
+
+      return combineLatest(allPages).pipe(
+        take(1),
+        map(contents => mapContentToSearchResult(contents, results)),
+      );
+    }
+    return of([]);
+  }
+
+  /** Triggers the reload of the search index from the remote. */
+  async triggerUpdate(): Promise<void> {
+    return await this._refreshSearchIndex();
+  }
+
+  /** Refresh the prebuilt search index. */
+  private async _refreshSearchIndex(): Promise<void> {
```

Comment:
please return stream don't cast to promise (to be consistent the find function uses streams as well)

```typescript
return this._http
      .get(`${environment.dataHost}/_search-index.json`)
      .pipe(tap(searchIndex => Index.load(searchIndex.data)))

```